src/domains/mapDomain.ml

```diff
@@ -149,7 +144,7 @@ struct
   let pretty_f short () mapping =
     let groups =
       let group_fold key itm gps =
-        let cl = Domain.classify key in
+        let cl = Domain.to_group key |> Domain.group_to_enum in
         match gps with
         | (a,n) when cl <>  n -> ((cl,(M.add key itm M.empty))::a, cl)
         | (a,_) -> ((fst (List.hd a),(M.add key itm (snd (List.hd a))))::(List.tl a),cl) in
```

Comment:
I tried to understand what it's doing, but it doesn't make sense as it is. The second case then has `n = cl` essentially. So the second component of the fold pair will always be the previous key's group. But then:
1. What about keys from alternating groups? Nothing about the groups is tied to the ordering (key `compare`) of the map, so it would only group consecutively?
2. Why is the second case consing a pair with `fst (List.hd a)` instead of just `cl` (the key's group)? How can it be anything other than `cl`?

Reply:
Yes, this code is way too convoluted and should be rewritten.
Is there some test that has enough grouped output that can be used to test it?

Reply:
Printing a `CPA` that contains a few local variables, global variables and function parameters should be observable from a fairly simple example, but I haven't actually tried. Also, I'm not sure how to get context and temp groups from a program, but for testing the grouping, a few of the above should be sufficient.

I've definitely seen them grouped in some tracing output. But I don't remember groups alternating, so maybe it currently handles proper grouping but I just didn't understand how.

Reply:
I just inserted a print in `MapDomain.LiftTop()().add` and it seems like it's working.
The `fold` traverses it in a fixed order such that the groups are consecutive. Maybe it relies on the keys' compare to ensure this.
The `group_to_enum` values don't affect the order - maybe it should be sorted by them afterwards?

Reply:
Rewrote it. Should work with non-consecutive groups and sorts by the variant type's order.

Reply:
While working on https://github.com/goblint/cil/issues/31 I think I figured out why the old implementation seemed to not have multiple non-consecutive copies of the same group in the output. The key ordering of `CPA` is based on `vid`s and by using `Partial.globally_unique_vids` those just happened to be renumbered in a way that all the parameters were numbered consecutively and then all the local variables were numbered consecutively etc.

Reply:
I just checked that indeed the old implementation doesn't group non-consecutively. For example here the global `main` was introduced by Goblint (after all the CIL stuff) for the return value of the thread (not the function) and ends up in a duplicate block:
```
mapping {
  Global {
    g -> (36)
  }
  Temp {
    RETURN -> (0)
  }
  Global {
    main -> (0)
  }
}
```

After this PR the globals are actually together which makes much more sense IMO.